lib/puppet/file_system/windows.rb

```diff
@@ -114,8 +118,136 @@ def read_preserve_line_endings(path)
+           else
+             raise ArgumentError, "#{mode} is invalid: Only modes 0644, 0640, 0660, 0666, 0600 and 0440 are allowed"
+           end
+
+    # puts "replace_file: dacl: #{dacl.inspect}"
+    tempfile = Puppet::FileSystem::Uniquefile.new(Puppet::FileSystem.basename_string(path), Puppet::FileSystem.dir_string(path))
+    begin
+      tempdacl = Puppet::Util::Windows::AccessControlList.new
+      tempdacl.allow(current_sid, FULL_CONTROL)
+      set_dacl(tempfile.path, tempdacl)
+      # puts "replace_file: tempfile 1 work"
+
+      begin
+        yield tempfile
+        tempfile.flush
+        tempfile.fsync
+      ensure
+        tempfile.close
+      end
+
+      set_dacl(tempfile.path, dacl) if dacl
+      File.rename(tempfile.path, Puppet::FileSystem.path_string(path))
+    ensure
+      tempfile.close!
+    end
+  rescue Puppet::Util::Windows::Error => e
+    case e.code
+    when ACCESS_DENIED, SHARING_VIOLATION, LOCK_VIOLATION
+      raise Errno::EACCES.new(Puppet::FileSystem.path_string(path), e)
+    else
+      raise SystemCallError.new(e.message)
+    end
+  end
+
   private
 
+  def set_dacl(path, dacl)
+    # Set the DACL
+    # This has a "special case" - if both Owner and Group are SYSTEM, then the Group field
+    # is set to BuiltinAdministrators (PUP-9719). This is needed to ensure that if Puppet Agent
+    # is run under SYSTEM/SYSTEM (e.g. under AWS services), that all internally managed Puppet files
+    # are are still writeable from the Administrator account
+    sd = Puppet::Util::Windows::Security.get_security_descriptor(path)
+    sd_group = sd.group == sd.owner && sd.group == Puppet::Util::Windows::SID::LocalSystem ? Puppet::Util::Windows::SID::BuiltinAdministrators : sd.group
+    # puts "set_dacl: sd.owner = #{sd.group}, sd.owner = #{sd.group}"
+    new_sd = Puppet::Util::Windows::SecurityDescriptor.new(sd.owner, sd_group, dacl, true)
+    Puppet::Util::Windows::Security.set_security_descriptor(path, new_sd)
+  end
+
+  def secure_dacl(current_sid, owner_permission = FULL_CONTROL, group_permission = FULL_CONTROL)
+    dacl = Puppet::Util::Windows::AccessControlList.new
+    [
+     Puppet::Util::Windows::SID::LocalSystem,
+     Puppet::Util::Windows::SID::BuiltinAdministrators,
+     current_sid
+    ].uniq.map do |sid|
+        permission = case sid
+                     when Puppet::Util::Windows::SID::LocalSystem
+                       owner_permission
+                     when Puppet::Util::Windows::SID::BuiltinAdministrators
+                       group_permission
+                     else
+                       owner_permission
+                     end
+        dacl.allow(sid, permission)
+    end
+    dacl
+  end
+
+  def get_dacl_from_file(path)
+    sd = Puppet::Util::Windows::Security.get_security_descriptor(Puppet::FileSystem.path_string(path))
+    sd.dacl
+  rescue Puppet::Util::Windows::Error => e
+    if e.code == 2 # ERROR_FILE_NOT_FOUND
```

Comment:
The error code could go as a constant under https://github.com/puppetlabs/puppet/pull/7594/files#diff-d56421f468d3cfe2988bff9fb021f192R121